rbd: provide option to disable setting metadata on rbd images

pkalever opened the pull request:
<!-- Please take a look at our [Contributing](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#Code-contribution-workflow)
documentation before submitting a Pull Request!
Thank you for contributing to ceph-csi! -->

# Describe what this PR does #

As we added support to set the metadata on the rbd images created for the PVC and volume snapshot, by default metadata is set on all the images.
                                                                          
As we have seen we are hitting issue#2327 a lot of times with this, we start to leave a lot of stale images. Currently, we rely on            
`--extra-create-metadata=true` to decide to set the metadata or not, we cannot set this option to false to disable setting metadata because we use this for encryption too.                                              
                                                                          
This changes is to provide an option to disable setting the image metadata when starting cephcsi.                                           
                                                                          

## Related issues ##

Fixes: #3009                                                              

Signed-off-by: Prasanna Kumar Kalever \<prasanna.kalever@redhat.com\>       

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

- `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)
- `/retest all`: run this in case the CentOS CI failed to start/report any test
  progress or results

</details>

Commit:
deploy: add setmetadata=true in the templates

setmetadata on the volume by default, otherwise e2e will fail

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

Commit:
doc: Add documentation about `--setmetadata` option

Fixes: #2874
Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

Commit:
rbd: provide option to disable setting metadata on rbd images

As we added support to set the metadata on the rbd images created for
the PVC and volume snapshot, by default metadata is set on all the images.

As we have seen we are hitting issues#2327 a lot of times with this,
we start to leave a lot of stale images. Currently, we rely on
`--extra-create-metadata=true` to decide to set the metadata or not,
we cannot set this option to false to disable setting metadata because we
use this for encryption too.

This changes is to provide an option to disable setting the image
metadata when starting cephcsi.

Fixes: #3009
Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

pkalever commented:
/retest all

Madhu-1 commented on charts/ceph-csi-rbd/values.yaml:
```diff
@@ -203,6 +203,9 @@ provisioner:
     resources: {}
 
+  # set metadata on volume
+  setmetadata: true
```
disable it by default until the ceph issue is fixed?

pkalever replied:
We should comment out the test cases then?

pkalever replied:
Ah! Noticed this now :-)
Still need to address it.

pkalever replied:
Done now.

Madhu-1 commented on charts/ceph-csi-rbd/values.yaml:
```diff
@@ -203,6 +203,9 @@ provisioner:
     resources: {}
 
+  # set metadata on volume
+  setmetadata: true
```
Need documentation update of helm parameters in Readme.md

Madhu-1 commented on cmd/cephcsi.go:
```diff
@@ -68,6 +68,11 @@ func init() {
 	flag.StringVar(&conf.NodeID, "nodeid", "", "node id")
 	flag.StringVar(&conf.PluginPath, "pluginpath", defaultPluginPath, "plugin path")
 	flag.StringVar(&conf.StagingPath, "stagingpath", defaultStagingPath, "staging path")
+	flag.BoolVar(
+		&conf.SetMetadata,
+		"setmetadata",
+		false,
+		"set metadata on the volume")
```
I think this won't exceed the line length limit. it can be a single statement like others,

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -1063,14 +1075,16 @@ func (cs *ControllerServer) CreateSnapshot(
+		// Update the metadata on snapshot not on the original image
+		rbdVol.RbdImageName = rbdSnap.RbdSnapName
+
+		err = rbdVol.setSnapshotMetadata(req.GetParameters())
```
can we make `SetMetadata` part of the rbdImage struct, if we do so we just need if check inside the  `setSnapshotMetadata` and `setVolumeMetadata`, and changes will be minimal and easy to remove this option later also? WDYT?

pkalever replied:
Sounds good to me.

Madhu-1 reviewed (dismissed):
few nits and questions.

nixpanic commented:
Just wondering, why a commandline option, and not a parameter in the StorageClass?

nixpanic commented on internal/controller/controller.go:
```diff
@@ -34,8 +34,9 @@ type Manager interface {
-	Namespace  string
+	DriverName  string
+	Namespace   string
+	SetMetadata bool // set metadata on the volume
```
comments are ideally above the attribute

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -1924,10 +1926,12 @@ func genVolFromVolIDWithMigration(
-		err := rv.SetMetadata(k, v)
-		if err != nil {
-			return fmt.Errorf("failed to set metadata key %q, value %q on image: %w", k, v, err)
+	if rv.EnableMetadata {
```
Do the reverse check, and return if setting the metadata is not needed. It is like testing for an error, and returning before executing further operations.

That reduces the indention for the next code block. It is the style we try to use throughout the project.

nixpanic reviewed (changes requested):
I'm not sure what the plan was with this PR. The option is deployment wide, whereas it could also be per StorageClass.

If I'm not mistaken, this is only required because there is a Ceph bug that gets hit regularly when the metadata is set on the RBD images? Link to the Ceph bug, and find out when a release includes a fix?

pkalever commented:
Rebased, solving many merge conflicts! I will test this manually and update the behavior shortly.

Commit:
cleanup: ReconcilePersistentVolume consider passing it by pointer

Address: hugeParam linter

internal/controller/persistentvolume/persistentvolume.go:59:7:
hugeParam: r is heavy (80 bytes); consider passing it by pointer
(gocritic)
[...]
internal/controller/persistentvolume/persistentvolume.go:135:7:
hugeParam: r is heavy (80 bytes); consider passing it by pointer
(gocritic)
func (r ReconcilePersistentVolume) reconcilePV(ctx context.Context, obj
runtime.Object) error {}

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

Commit:
cleanup: reduce struct padding

internal/rbd/rbd_util.go:89:15: struct of size 312 bytes could be of
size 304 bytes:
``
struct{
	RbdImageName   	string,
	ImageID        	string,
	VolID          	string,
	Monitors       	string,
	JournalPool    	string,
	Pool           	string,
	RadosNamespace 	string,
	ClusterID      	string,
	RequestName    	string,
	NamePrefix     	string,
	ParentName     	string,
	ParentPool     	string,
	ClusterName    	string,
	Owner          	string,
	VolSize        	int64,
	StripeCount    	uint64,
	StripeUnit     	uint64,
	ObjectSize     	uint64,
	ImageFeatureSet	github.com/ceph/go-ceph/rbd.FeatureSet,
	encryption
*github.com/ceph/ceph-csi/internal/util.VolumeEncryption,
	CreatedAt
*google.golang.org/protobuf/types/known/timestamppb.Timestamp,
	conn
*github.com/ceph/ceph-csi/internal/util.ClusterConnection,
	ioctx          	*github.com/ceph/go-ceph/rados.IOContext,
	Primary        	bool,
	EnableMetadata 	bool,
}
`` (maligned)
type rbdImage struct {
              ^}`
make: *** [Makefile:118: go-lint] Error 1

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

pkalever commented:
@Madhu-1 I got the same struct padding error now, see https://github.com/ceph/ceph-csi/runs/6916598994?check_suite_focus=true

pkalever commented:
> Rebased, solving many merge conflicts! I will test this manually and update the behavior shortly.

Testing is done, and the changes are working as expected.

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.21

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.21

pkalever commented:
@mergifyio rebase

pkalever commented:
@Mergifyio rebase

pkalever commented:
/retest all

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.22

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.23

nixpanic commented:
@mergifyio rebase

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.21

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.22

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.23

pkalever commented:
[Deployment failure
](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.21/detail/mini-e2e-helm_k8s-1.21/4344/pipeline)/retest ci/centos/mini-e2e-helm/k8s-1.21

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23
```
Jun 21 10:32:16.449: FAIL: backend images not matching kubernetes resource count,image count 22 kubernetes resource count 21
```
[logs](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/rest/organizations/jenkins/pipelines/mini-e2e-helm_k8s-1.23/runs/2188/nodes/107/log/?start=0)

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.21

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.22

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.23

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.21

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.21

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.21

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.21

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.21

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.21

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -2061,6 +2059,10 @@ func genVolFromVolIDWithMigration(
 
 // setAllMetadata set all the metadata from arg parameters on RBD image.
 func (rv *rbdVolume) setAllMetadata(parameters map[string]string) error {
+	if !rv.EnableMetadata {
```
@pkalever same check is also required in unsetMetadata to skip RemoveMetadata call or else we might end up calling removemetadata and discard not found erro?

pkalever replied:
I had intentionally not mentioned that in the RemoveMetadata, because if the user had enabled setmetadata at some point in time and disabled it later, the calls coming to RemoveMetadata should be handled and metadata should be removed on previous images (in case metadata exists on them)?

Madhu-1 replied:
we can document the expectations WDYT?. adding check-in unsetMetadata helps in avoiding 3 calls RemoveMetadat call.

pkalever replied:
IMHO, it is our csi responsibility to clean the metadata that it has added. Moreover, in case metadata is absent on the image, it will skip the rest of the operation to remove metadata. But that all said, I would like to leave the final decision to you!

Madhu-1 replied:
@nixpanic any suggestion?

nixpanic replied:
re-considered this approach, and I think it is cleanest if the caller decides if the metadata needs to be set. The `EnableMetadata` is not really an option for the `rbdImage`, but rather the Controller or Node-Plugin.

If the option is changed after volumes have been created with metadata, keeping stale metadata is acceptable, it should just not be used.

pkalever replied:
Cool, I will consider adding `if !rv.EnableMetadata{}` defence check in removeMetadata() too, Thanks guys!

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.21

pkalever commented:
@nixpanic @Madhu-1 @Rakshith-R please replace your acks!

**Diff**
```
[0] pkalever 😎 ceph-csi✨ git diff     
diff --git a/internal/rbd/rbd_util.go b/internal/rbd/rbd_util.go
index e8e1f40a9..59c26f187 100644          
--- a/internal/rbd/rbd_util.go              
+++ b/internal/rbd/rbd_util.go
@@ -2087,6 +2087,10 @@ func (rv *rbdVolume) setAllMetadata(parameters map[string]string) error {
                                       
 // unsetAllMetadata unset all the metadata from arg keys on RBD image.
 func (rv *rbdVolume) unsetAllMetadata(keys []string) error {
+       if !rv.EnableMetadata {                       
+               return nil                                                                                             
+       }                              
+                                                                                                                      
        for _, key := range keys {
                err := rv.RemoveMetadata(key)
                // TODO: replace string comparison with errno.
```

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.21

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.21

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.22

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.23

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.21

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.21

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.21

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.23

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.23

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.21

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.23

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.23

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.21

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.23

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.21

Madhu-1 commented:
@mergifyio rebase

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.22

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.21

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.21

pkalever commented:
@Madhu-1 @nixpanic @ceph/ceph-csi-contributors 
Seeing this, not sure why!
```
[1mSTEP[0m: create a PVC and check PVC/PV metadata on RBD image
[...]
Jun 24 04:51:27.094: INFO: stdErr occurred: failed to get metadata csi.storage.k8s.io/pvc/name of image : (2) No such file or directory
rbd: getting metadata failed: (2) No such file or directory

Jun 24 04:51:27.094: FAIL: failed to get PVC name --pool=replicapool/csi-vol-ea2d0eb1-f370-11ec-8161-0ed2bb196630 csi.storage.k8s.io/pvc/name: err=command terminated with exit code 2 stdErr="failed to get metadata csi.storage.k8s.io/pvc/name of image : (2) No such file or directory\nrbd: getting metadata failed: (2) No such file or directory\n"
[AfterEach] RBD
```
[logs](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/rest/organizations/jenkins/pipelines/mini-e2e_k8s-1.23/runs/2081/nodes/89/log/?start=0)

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.22

pkalever commented:
@ceph/ceph-csi-contributors Here is the issue to track the e2e changes https://github.com/ceph/ceph-csi/issues/3220

Madhu-1 commented on charts/ceph-csi-rbd/README.md:
```diff
@@ -112,6 +112,7 @@ charts and their default values.
 | `provisioner.skipForceFlatten`                 | Skip image flattening if kernel support mapping of rbd images which has the deep-flatten feature                                                     | `false`                                            |
 | `provisioner.timeout`                          | GRPC timeout for waiting for creation or deletion of a volume                                                                                        | `60s`                                              |
 | `provisioner.clustername`                      | Cluster name to set on the RBD image                                                                                                                 | ""                                                 |
+| `provisioner.setmetadata`                      | Set metadata on volume                                                                                                                               | `false`                                            |
```
default is true in helm charts

pkalever replied:
@Madhu-1 this is addressed now, thanks!

Rakshith-R replied:
it should contain internal default that is false ? 

	flag.BoolVar(&conf.SetMetadata, "setmetadata", false, "set metadata on the volume")

pkalever replied:
Hmm, right!

pkalever replied:
Sorry got confused, I think its values from charts!

Madhu-1 replied:
AFAIK this is the documentation of the default values in values.yaml.

Rakshith-R replied:
I believe it should internal default value since this will be used when no value in set for a variable in values.yaml.

pkalever replied:
Lets come to a conclusion, we need this PR merged ASAP.

Madhu-1 commented:
Removed backport label please feel free to add it back if its required.

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.22

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.21

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

nixpanic commented:
> Removed backport label please feel free to add it back if its required.

This is a new feature,  so it should not get backported.

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.21

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

nixpanic commented:
This can be rebased with Mergify once #3216 is merged (that one is 1st in the queue atm).

pkalever commented:
@Mergifyio rebase

nixpanic commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

nixpanic commented:
> /retest ci/centos/mini-e2e-helm/k8s-1.22

```
Jun 28 15:51:43.874: FAIL: backend images not matching kubernetes resource count,image count 12 kubernetes resource count 11
```
[logs](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.22/detail/mini-e2e-helm_k8s-1.22/3648/pipeline)

nixpanic commented:
/retest ci/centos/k8s-e2e-external-storage/1.23

nixpanic commented:
> /retest ci/centos/k8s-e2e-external-storage/1.23

```
[Fail] External Storage [Driver: cephfs.csi.ceph.com] [Testpattern: Dynamic PV (default fs)] subPath [It] should support existing directories when readOnly specified in the volumeSource 
```
[logs](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/k8s-e2e-external-storage-1.23/detail/k8s-e2e-external-storage-1.23/2050/pipeline)
